django_ses/models.py

```diff
@@ -14,3 +14,10 @@ class Meta:
 
     def __str__(self):
         return self.date.strftime("%Y-%m-%d")
+
+
+class BlacklistedEmail(models.Model):
+    email = models.EmailField(max_length=255, unique=True)
```

Comment:
This is case sensitive FYI.

https://code.djangoproject.com/ticket/17561

Maybe we should add a separate setting to force-lowercase emails, some people may want to store them case-sensitive.

`AWS_SES_BLACKLIST_FORCE_LOWERCASE` (or similar). Default `False`. In this case, don't apply `.lower()` anywhere, including in the management command.

Otherwise, use current logic.

Reply:
I'm aware that the Email field is case sensitive. The `bounce_handler` and the `complaint_handler` store the emails in lowercase (`.lower()`) so the `get_or_create()` and the `filter_blacklisted_recipients()` can assume that the emails in the database are lowercased and perform the db query with lowercased emails. (more on that in a bit)

I assume that the emails **must** be case insensitive. It's true that, according to RFC 1035, emails **could** be case sensitive, but there is virtually no mailing system (not at least that I'm aware of) that has ever supported case sensitivity in the local part of the email (and imho supporting that is just asking for all kind of problems).

(more on that) We **could** change the code and store the emails in a case sensitive way, but there are some implications in doing that:

a) The `email` field is created with `unique=True`, which implies that `text_pattern_ops` index is created on that field (in PostgreSQL) (because of https://github.com/django/django/commit/f1ea26dd99). That type of index (https://www.postgresql.org/docs/9.3/indexes-opclass.html) doesn't support case insensitive search. The `filter_blacklisted_recipients()` could do `BlacklistedEmail.objects.filter(email__iexact=email)`, but that will trigger a FULL INDEX search (very slow and expensive).

b) We could fix `a)` by using `db_collation` but, as the docs state (https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.CharField.db_collation), collation names are not standardised, which means that we'd need to either:

b.1) force only PostgreSQL on the extension for compatibility reasons
b.2) research the right db_collation for each database (how many do we want to support?) and hardcode it in the code (and somehow pick the right collation when the migration is executed)

IMHO, an extremely high amount of work / effort in supporting a feature that (again, imho), nobody will miss. Case insensitivity in emails just doesn't exist anywhere in practice, and supporting it just covers some theoretical edge cases.

That said, you have the final word 🙏. If you want me to, I can try and implement this.

Reply:
Having second thoughts on this one... 🤔 We could implement it the way you said (`AWS_SES_BLACKLIST_FORCE_LOWERCASE `), but the default could be `True` (emails get lowercased by default). We can then include a warning in the README about setting this to `False`: we won't do anything special in the `get_or_create()` and `filter_blacklisted_recipients()` (`john@foo.com` and `John@foo.com` could get inserted in the blacklist, sending a mail to `johN@foo.com` won't be filtered). The user will be aware that the emails **will** be threaded in a case sensitive way.

Reply:
@alexandernst thank you for the thoughtful reply here.

I think it makes sense to give people the option to have case-insensitivity, but going back to what you said... probably nobody will use it.

So, I am OK to keep the code as you have it, only supporting lowercase emails, but make sure to document that the blacklist feature does not support case-sensitive emails. For those who want case-sensitive blacklisting, they can file a feature request or implement their own.

Alternatively if you want to add the flag, that's fine, but it probably adds complexity that will never be used.